pkg/controller/zookeepercluster/zookeepercluster_controller.go

```diff
@@ -295,7 +298,31 @@ func (r *ReconcileZookeeperCluster) reconcileStatefulSet(instance *zookeeperv1be
 
 			data := "CLUSTER_SIZE=" + strconv.Itoa(int(newSTSSize))
 			r.log.Info("Updating Cluster Size.", "New Data:", data, "Version", version)
-			r.zkClient.UpdateNode(path, data, version)
+			err = r.zkClient.UpdateNode(path, data, version)
+			if err != nil {
+				return fmt.Errorf("Error updating cluster size %s: %v", path, err)
+			}
+			// #398 if decrease node, remove node immediately after updating node successfully.
+			if newSTSSize < foundSTSSize {
+				var removes []string
+				config, _, err := r.zkClient.GetConfig()
+				if err != nil {
+					return fmt.Errorf("Error GetConfig %v", err)
+				}
+				r.log.Info("Get zookeeper config.", "Config: ", config)
+				for myid := newSTSSize + 1; myid <= foundSTSSize; myid++ {
+					if strings.Contains(config, "server."+strconv.Itoa(int(myid))+"=") {
+						removes = append(removes, strconv.Itoa(int(myid)))
+					}
+				}
+				// The node that have been removed with reconfig also can still provide services for all online clients.
+				// So We can remove it firstly, it will avoid to error that client can't connect to server on preStop.
+				r.log.Info("Do reconfig to remove node.", "Remove ids", strings.Join(removes, ","))
```

Comment:
we are doing the same operation of removing node in https://github.com/pravega/zookeeper-operator/blob/master/docker/bin/zookeeperTeardown.sh#L45. Do you still think removing node is required here?

Reply:
@stop-coding Did you see `Do reconfig to remove node.` message being present in the logs in your use case?

I think @anishakj is suggesting that catching the `UpdateNode` error and returning on line 303 should be enough to fix the issue, hence lines 305 to 324 would never get executed.

Reply:
@jkhalack  Sorry for my late reply.
Only  catching the  _UpdateNode error_ is not enough to ensure do reconfigure successfully on _preStop_, for example that pod exiting but the cluster broken again. We indeed hope that updating node size and doing reconfigure is atomicity, but it's not realistic. 

It known that updating "Spec.Replicas" to k8s will tell pod to create or exit. If it fail on scale down, we can stop updating "Spec.Replicas"  until cluster recovery, that will ensure have done reconfigure before pod exit.

So I think that doing reconfigure on checking the cluster scale down is better.